### PR TITLE
ARSN-242 Fix non-current version transition

### PR DESCRIPTION
**Pull request by nicolas2bert**

### CODE REVIEW

File: lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts [hunk 1]
```diff
@@ -239,6 +316,16 @@ export default class LifecycleUtils {
                     currentDate,
                 });
             }
+
+            const hasNoncurrentVersionTransitions = Array.isArray(rule.NoncurrentVersionTransitions) && rule.NoncurrentVersionTransitions.length > 0;
+            if (hasNoncurrentVersionTransitions && this._supportedRules.includes('noncurrentVersionExpiration')) {
```
Comment (rachedbenmustapha) [inline lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts]: should the supported rule be `noncurrentVersionTransition` ?

File: lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts [hunk 2]
```diff
@@ -98,6 +139,42 @@ export default class LifecycleUtils {
         });
     }
 
+    /**
+    * Find the most relevant non-current version trantition rule for the given transitions array
```
Comment (rachedbenmustapha) [inline lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts]:
```suggestion
    * Find the most relevant non-current version transition rule for the given transitions array
```

File: lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts [hunk 3]
```diff
@@ -98,6 +139,42 @@ export default class LifecycleUtils {
...
+    * and any previously stored non-current version transition from another rule.
+    * @param params - The function parameters
+    * @param params.transitions - Array of lifecycle non-current version transitions
+    * @param params.lastModified - The object's last modified
+    * date
+    * @return The most applicable non-current version transition rule
+    */
+    getApplicableNCVTransition(params: {
```
Comment (rachedbenmustapha) [inline lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts:151]: More of a question: given the new constraint that we can't transition out of cold storage (yet?) is it something that should be handled here, or at the call site? Call site seems to make more sense as the backbeat bucket processor has access to the location's `isCold` attribute.

File: lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts (hunk 1 above)
Comment (nicolas2bert) [reply lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts]: Yes, thanks!

File: lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts (hunk 3 above)
Comment (nicolas2bert) [reply lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts:151]: Yes, I think "backbeat bucket processor" is a good place to handle it. 

Also, we can't replication an object located in a Cold storage location. We will need to handle this limitation as well. 

Note: We already handle replicating "to" a Cold storage location but not "from" there.

Comment (nicolas2bert) [reply lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts:151]: I created a ticket for it: https://scality.atlassian.net/browse/BB-241

**Review by rachedbenmustapha — APPROVED** [review]

**Review by miniscruff — APPROVED** [review]

**Comment by nicolas2bert** [comment]
@bert-e approve

**Pull request merged**